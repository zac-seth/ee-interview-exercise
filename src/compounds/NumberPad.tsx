import { FC, useCallback, useState } from "react";

import {
  Grid,
  HandleValueSelected,
  NumberButton,
  PlaceholderButton,
} from "elements";

type Digit = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

interface HandleProvideDigit {
  (digit: Digit): void;
}

export type NumberPadProps = {
  onValueSelected: HandleValueSelected;
};

type ValueItem = { value: Digit };
type PlaceholderItem = { isPlaceholder: true };
type Item = ValueItem | PlaceholderItem;

const items: Item[] = [
  { value: 7 },
  { value: 8 },
  { value: 9 },
  { value: 4 },
  { value: 5 },
  { value: 6 },
  { value: 1 },
  { value: 2 },
  { value: 3 },
  { isPlaceholder: true },
  { value: 0 },
  { isPlaceholder: true }
];

const isValue = (item: Item): item is ValueItem => item.hasOwnProperty(`value`);

export const NumberPad: FC<NumberPadProps> = ({ onValueSelected }) => {
  const [value, setValue] = useState(0);
  const handleProvideDigit = useCallback<HandleProvideDigit>(
    digit => {
      const newValue = value === 0 ? digit : +`${value}${digit}`;

      setValue(newValue);
      onValueSelected(newValue);
    },
    [onValueSelected, value]
  );

  return (
    <Grid cols={3}>
      {items.map((item, i) =>
        isValue(item) ? (
          <NumberButton
            label={item.value.toString()}
            key={`item-${i}`}
            onClick={handleProvideDigit}
            value={item.value}
          />
        ) : (
          <PlaceholderButton label="?" key={`item-${i}`} />
        )
      )}
    </Grid>
  );
};
