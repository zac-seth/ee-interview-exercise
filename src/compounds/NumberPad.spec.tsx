import { fireEvent, render, screen } from "@testing-library/react";

import { NumberPad } from "./NumberPad";

const handleValueSelected = jest.fn();

describe(`NumberPad`, () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it(`Should render the number pad`, () => {
    render(<NumberPad onValueSelected={handleValueSelected} />);

    expect(screen.getByTestId(`grid-layout`)).toMatchSnapshot();
  });

  it(`Should render the correct buttons in the pad`, async () => {
    render(<NumberPad onValueSelected={handleValueSelected} />);

    const buttons = await screen.findAllByRole(`button`);

    expect(buttons.length).toBe(12);
  });

  it(`Should correctly pass the value selected`, async () => {
    render(<NumberPad onValueSelected={handleValueSelected} />);

    fireEvent.click(screen.getByText(`1`));
    fireEvent.click(screen.getByText(`2`));
    fireEvent.click(screen.getByText(`3`));

    expect(handleValueSelected).toHaveBeenCalledWith(1);
    expect(handleValueSelected).toHaveBeenCalledWith(12);
    expect(handleValueSelected).toHaveBeenCalledWith(123);
  });
});
