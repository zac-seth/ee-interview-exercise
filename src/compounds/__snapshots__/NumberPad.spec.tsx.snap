// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberPad Should render the number pad 1`] = `
.c1 {
  font-family: 'Lato',sans-serif;
  font-weight: 400;
  color: #666;
  text-shadow: 0 0 5px #1A94D2,0 0 15px #228DFF;
  font-size: 4rem;
  line-height: 4rem;
  text-align: center;
  padding: 1rem;
  margin: 0;
  width: 8rem;
  height: 8rem;
  background-color: inherit;
  color: #1a94d2;
  border: 0;
  cursor: pointer;
}

.c1:hover {
  background-color: #1a1a1a;
}

.c2 {
  font-family: 'Lato',sans-serif;
  font-weight: 400;
  color: #666;
  text-shadow: 0 0 5px #1A94D2,0 0 15px #228DFF;
  font-size: 4rem;
  line-height: 4rem;
  text-align: center;
  padding: 1rem;
  margin: 0;
  width: 8rem;
  height: 8rem;
  background-color: inherit;
  color: #1a94d2;
  border: 0;
  cursor: pointer;
  color: #666;
}

.c2:hover {
  background-color: #1a1a1a;
}

.c2:hover {
  background-color: inherit;
  cursor: default;
}

.c0 {
  display: grid;
  grid-gap: 0.25rem;
  margin: 0;
  padding: 0;
  grid-template-columns: repeat(3,1fr);
}

<div
  class="c0"
  data-testid="grid-layout"
>
  <button
    class="c1"
  >
    7
  </button>
  <button
    class="c1"
  >
    8
  </button>
  <button
    class="c1"
  >
    9
  </button>
  <button
    class="c1"
  >
    4
  </button>
  <button
    class="c1"
  >
    5
  </button>
  <button
    class="c1"
  >
    6
  </button>
  <button
    class="c1"
  >
    1
  </button>
  <button
    class="c1"
  >
    2
  </button>
  <button
    class="c1"
  >
    3
  </button>
  <button
    class="c2"
  >
    ?
  </button>
  <button
    class="c1"
  >
    0
  </button>
  <button
    class="c2"
  >
    ?
  </button>
</div>
`;
